### PR TITLE
small logging changes to train, more accurate ppl

### PR DESCRIPTION
**Pull request by alexholdenmiller**
- perplexity is now calculated using average of the average loss (this is not fully accurate but better than current)
- train_model now tells you what the previous best valid is when a new best is found
- documentation for personachat options

### CODE REVIEW

**Comment by alexholdenmiller** [comment]
old ppl was sensitive to batch size, new one still has a slight sensitivity but much less.

ran personachat:self eval on a given train model, my_ppl is the new calculation:
```
bs 1   'ppl': 155.8, 'my_ppl': 66.09
bs 2   'ppl': 88.89, 'my_ppl': 63.43
bs 8   'ppl': 69.31, 'my_ppl': 64.01
bs 32  'ppl': 66.01, 'my_ppl': 64.33
bs 64  'ppl': 65.45, 'my_ppl': 64.33
bs 128 'ppl': 65.09, 'my_ppl': 64.33
bs 128 'ppl': 65.09, 'my_ppl': 64.33
bs 256 'ppl': 65.05, 'my_ppl': 64.39
bs 512 'ppl': 65.11, 'my_ppl': 64.59
bs 512 'ppl': 65.11, 'my_ppl': 64.59
```

File: parlai/core/metrics.py [hunk 1]
```diff
@@ -258,6 +258,9 @@ def report(self):
             for k in self.metrics_list:
                 if self.metrics[k + '_cnt'] > 0 and k != 'correct' and k != 'f1':
                     m[k] = round_sigfigs(self.metrics[k] / self.metrics[k + '_cnt'], 4)
+        # if 'loss' in m and 'ppl' not in m:
+        if 'loss' in m:
+            m['ppl'] = round_sigfigs(math.exp(m['loss']), 4)
```
Comment (jaseweston) [inline parlai/core/metrics.py]: this doesn't make sense in that loss could mean anything (0/1, ranking, cross ent, mse).. this should only be for a specific kind of loss

Comment (alexholdenmiller) [reply parlai/core/metrics.py]: yeah I was a little worried about that too, hmm I think I found a better way to do this

**Comment by alexholdenmiller** [comment]
quick note: will fix #609

**Comment by alexholdenmiller** [comment]
New perplexity results are much more consistent:
```
1:    'loss': 4.218, 'ppl': 67.89
2:    'loss': 4.167, 'ppl': 64.55
512:  'loss': 4.167, 'ppl': 64.55
```

There's a small difference with bs = 1 which I currently believe is due to rounding errors in the decimals. Still trying to check.

File: parlai/agents/language_model/language_model.py [hunk 2]
```diff
@@ -403,8 +425,8 @@ def predict(self, data, hidden, targets=None, is_training=True, y_lens=None):
             if bsz != self.batchsize:
                 self.hidden = self.model.init_hidden(bsz)
             loss = self.get_target_loss(data, self.hidden, targets, y_lens)
-            loss_dict['loss'] = loss
-            loss_dict['ppl'] = math.exp(loss)
+            self.metrics['loss'] += loss
+            self.metrics['num_tokens'] += sum(y_lens)
 
         return output, hidden, loss_dict, predictions
```
Comment (emilydinan) [inline parlai/agents/language_model/language_model.py]: seems like you can just get rid of `loss_dict` and not return it since it will always return `None`. You should also get rid of all of the `loss_dict` code in the `batch_act` function

File: parlai/core/worlds.py [hunk 3]
```diff
@@ -359,12 +367,21 @@ def episode_done(self):
...
+        for a in self.agents:
+            if hasattr(a, 'report'):
+                m = a.report()
+                for k, v in m.items():
+                    if k not in metrics:
+                        # first agent gets priority in settings values for keys
+                        # this way model can't e.g. override accuracy to 100%
+                        metrics[k] = v
+        if metrics:
+            if compute_time and 'total' in metrics:
```
Comment (emilydinan) [inline parlai/core/worlds.py]: are there cases when we do not count the total?

Comment (alexholdenmiller) [reply parlai/core/worlds.py]: eh not that I can think of but seems reasonable to check

File: parlai/agents/language_model/language_model.py (hunk 2 above)
Comment (alexholdenmiller) [reply parlai/agents/language_model/language_model.py]: good catch

**Comment by alexholdenmiller** [comment]
confirmed that the bs = 1 is not rounding error in summing losses, will continue to investigate

File: examples/train_model.py [hunk 4]
```diff
@@ -101,8 +101,7 @@ def run_eval(agent, opt, datatype, max_exs=-1, write_log=False, valid_world=None
 
     if valid_world is None:
         valid_world = create_task(opt, agent)
-    else:
-        valid_world.reset()
+    valid_world.reset()
```
Comment (jaseweston) [inline examples/train_model.py]: is this the fix we were talking about before?

**Comment by alexholdenmiller** [comment]
Update: got completely consistent results training s2s on babi:task10k:1.

Trained model with
```
py examples/train_model.py -t babi:task10k:1 --dict-file /tmp/dict_babi:task10k:1 -bs 32 -vtim 30 -vcut 0.95 -m seq2seq -mf /tmp/testbabi
```

Eval with two modes: regular 1000-ex eval, then do eval multitasking on the same data over and over.
```
py examples/eval_model.py -t babi:task10k:1{,babi:task10k:1...} --dict-file /tmp/dict_babi:task10k:1 -m seq2seq -mf /tmp/testbabi --gpu 1 -dt {mode} -bs {bsz}
```

Results:
```
valid1k
bs 1    'loss': 0.02498, 'ppl': 1.025
bs 1024 'loss': 0.02498, 'ppl': 1.025

test1k
bs 1    'loss': 0.02205, 'ppl': 1.022
bs 1024 'loss': 0.02205, 'ppl': 1.022

valid 10k
bs 1    'loss': 0.02498, 'ppl': 1.025
bs 1024 'loss': 0.02498, 'ppl': 1.025

test10k
bs    1 'loss': 0.02205, 'ppl': 1.022
bs 1024 'loss': 0.02205, 'ppl': 1.022

valid50k
bs 1    'loss': 0.02498, 'ppl': 1.025
bs 1024 'loss': 0.02498, 'ppl': 1.025
```

So this is looking good to me, let's land.

File: examples/train_model.py (hunk 4 above)
Comment (alexholdenmiller) [reply examples/train_model.py]: this makes it so that the agent forgets any partial observations of training examples going into validation

Comment (jaseweston) [reply examples/train_model.py]: cool, yea that was the known bug that's in the TO DO, right?

Comment (alexholdenmiller) [reply examples/train_model.py]: that's right! the custom metrics made it really obvious that there was carryover, because e.g. the lmloss in the language model (training loss) was being reported during the validation, and vice-versa. so these resets on both sides keep things clean, since the metrics are also reset.

Comment (alexholdenmiller) [reply examples/train_model.py]: validation should finish episodes, though this does mean that training may have some examples truncated. that doesn't seem like a huge cost unless you're validating like every dozen examples though...
